ang/civicase/DashboardTab.js

```diff
@@ -1,10 +1,109 @@
-(function (angular) {
+(function (angular, $, _) {
   var module = angular.module('civicase');
 
   module.directive('civicaseDashboardTab', function () {
     return {
       restrict: 'E',
+      controller: 'dashboardTabCtrl',
       templateUrl: '~/civicase/DashboardTab.html'
     };
   });
-})(angular);
+
+  module.controller('dashboardTabCtrl', dashboardTabCtrl);
+
+  dashboardTabCtrl.$inject = ['$location', '$scope', 'ContactsDataService', 'formatCase'];
+
+  function dashboardTabCtrl ($location, $scope, ContactsDataService, formatCase) {
+    var CASES_QUERY_PARAMS_DEFAULTS = {
+      'status_id.grouping': 'Opened',
+      'options': { 'sort': 'start_date DESC' }
+    };
+
+    $scope.newCasesPanel = {
+      query: {
+        entity: 'Case',
+        action: 'getcaselist',
+        params: casesQueryParams()
+      },
+      handlers: {
+        range: casesRangeHandler,
+        results: casesResultsHandler
+      },
+      custom: {
+        itemName: 'cases',
+        caseClick: casesCustomClick
+      }
+    };
+
+    (function init () {
+      initWatchers();
+    }());
+
+    /**
+     * Click handler that redirects the browser to the given case's details page
+     *
+     * @param {Object} theCase
+     */
+    function casesCustomClick (theCase) {
+      $location.path('case/list').search('caseId', theCase.id);
+    }
+
+    /**
+     * Returns a copy of the default query params for cases, interpolated with
+     * the currently selected relationship type filter's params
+     *
+     * @return {Object}
+     */
+    function casesQueryParams () {
+      return _.assign({}, CASES_QUERY_PARAMS_DEFAULTS, $scope.activityFilters.case_filter);
+    }
+
+    /**
+     * Sets the range of the case start date
+     *
+     * @param {String} selectedRange the currently selected period range
+     * @param {Object} queryParams
+     */
+    function casesRangeHandler (selectedRange, queryParams) {
+      var now = moment();
+      var end = now.endOf(selectedRange).format('YYYY-MM-DD');
+      var start = now.startOf(selectedRange).format('YYYY-MM-DD');
+
+      queryParams.start_date = { 'BETWEEN': [start, end] };
+    }
+
+    /**
+     * Formats each cases returned by the api call, and fetches the data
+     * of all the contacts referenced in every case
+     *
+     * @param {Array} results
+     */
+    function casesResultsHandler (results) {
+      // Flattened list of all the contact ids of all the contacts of all the cases
+      var contactIds = _(results).pluck('contacts').flatten().pluck('contact_id').uniq().value();
+      var formattedResults = results.map(formatCase);
+
+      // The try/catch block is necessary because the service does not
+      // return a Promise if it doesn't find any new contacts to fetch
+      try {
+        return ContactsDataService.add(contactIds)
```

Comment:
Can we just return `$q.resolve()` from `ContactsDataService.add`?

Reply:
Given our timeline and my low familiarity with the codebase, i really don't want to change upstream services if i don't absolutely have to, for fear of regression issues